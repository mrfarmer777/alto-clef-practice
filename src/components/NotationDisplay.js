import { Voice, Stave, StaveNote, Formatter, Renderer } from 'vexflow';
import { useEffect } from 'react'

function NotationDisplay(props){

  const buildNote = () => {
    const noteName = props.targetNote ?? "c"
    const octave = props.octave ?? "4"
    return new StaveNote( { clef: "alto", keys: [`${noteName}/${octave}`], duration: "w"})
  }

  const numberToHex = (value) => {
    const number = Math.floor(0.01 * value * 255);
    return `#${number.toString(16)}${number.toString(16)}${number.toString(16)}`
  };

  useEffect(() => {
    const outputElement = document.getElementById('output');
    outputElement.innerHTML = '';
    const renderer = new Renderer(outputElement, Renderer.Backends.SVG);
    renderer.resize(400,275);

    const context = renderer.getContext();
    context.scale(2,2);
    context.setFont('Arial', 10);

    const formatter = new Formatter();
    const trebleStave = new Stave(0, -15, 400);
    trebleStave.addClef('treble').setStyle({strokeStyle: numberToHex(props.opacity), fillStyle: numberToHex(props.opacity)})

    const bassStave = new Stave(0, 45, 400);
    bassStave.addClef('bass').setStyle({strokeStyle: numberToHex(props.opacity), fillStyle: numberToHex(props.opacity)})

    const altoStave = new Stave(70,15, 350);
    altoStave.addClef('alto')

    const targetNote = buildNote();
    const voice = new Voice({num_beats: 4, beat_value: 4})
    voice.addTickables([targetNote])


    formatter.joinVoices([voice]).format([voice])

    trebleStave.setContext(context).drawWithStyle();
    bassStave.setContext(context).drawWithStyle();
    context.setStrokeStyle('#000')
    voice.setStave(altoStave);
    altoStave.setContext(context).draw();
    voice.draw(context);

    const trebleTickables = []
    if(props.trebleHelpers.length > 0){
      trebleTickables.push(new StaveNote({ clef: 'treble', keys: props.trebleHelpers, duration: "w" }))
      const trebleVoice = new Voice({num_beats: 4, beat_value: 4})
      trebleVoice.addTickables(trebleTickables)
      trebleVoice.setStave(trebleStave).setStyle({strokeStyle: numberToHex(props.opacity), fillStyle: numberToHex(props.opacity)});
      formatter.joinVoices([trebleVoice]).format([trebleVoice])
      trebleVoice.setContext(context).drawWithStyle();
    }

    const bassTickables = []
    if(props.bassHelpers.length > 0){
      bassTickables.push(new StaveNote({ clef: 'bass', keys: props.bassHelpers, duration: "w" }))
      const bassVoice = new Voice({num_beats: 4, beat_value: 4})
      bassVoice.addTickables(bassTickables)
      bassVoice.setStave(bassStave).setStyle({strokeStyle: numberToHex(props.opacity), fillStyle: numberToHex(props.opacity)});;
      formatter.joinVoices([bassVoice]).format([bassVoice])
      bassVoice.setContext(context).drawWithStyle();
    }
  })


  return ('')
}

export default NotationDisplay;