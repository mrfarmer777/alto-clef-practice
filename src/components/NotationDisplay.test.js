import { render, screen } from '@testing-library/react';
import NotationDisplay from './NotationDisplay';

test('renders an svg canvas', () => {
  render(
    <div>
      <NotationDisplay />
      <div id={'output'} title={'notation-output'}></div>
    </div>
    );
  const content = screen.getByTitle(/notation-output/i);

  expect(content).toBeInTheDocument();
  const svgEl = content.querySelector("[viewBox='0 0 500 500']")
  expect(svgEl).toBeDefined();
});

test('it renders the note given by the prop', ()=>{
  render(
    <div>
      <NotationDisplay targetNote='c' octave="3" />
      <div id={'output'} title={'notation-output'}></div>
    </div>
  )

  const content = screen.getByTitle(/notation-output/i);
  const originalSvg = content.querySelector("[viewBox='0 0 500 500']")
  const originalNoteHeadEl = originalSvg.querySelector(".vf-notehead path")

  expect(originalNoteHeadEl).toMatchSnapshot();

  render(
    <div>
      <NotationDisplay targetNote='d' octave="3"/>
    </div>
  )

  const newContent = screen.getByTitle(/notation-output/i);
  const secondSvg = newContent.querySelector("[viewBox='0 0 500 500']")
  const noteHeadEl = secondSvg.querySelector(".vf-notehead path")
  expect(noteHeadEl).toMatchSnapshot()
})

test('it renders the note in the octave given by the prop', ()=>{
  render(
    <div>
      <NotationDisplay targetNote='d' octave='4' />
      <div id={'output'} title={'notation-output'}></div>
    </div>
  )

  const newContent = screen.getByTitle(/notation-output/i);
  const secondSvg = newContent.querySelector("[viewBox='0 0 500 500']")
  const noteHeadEl = secondSvg.querySelector(".vf-notehead path")

  expect(noteHeadEl).toBeDefined();
})