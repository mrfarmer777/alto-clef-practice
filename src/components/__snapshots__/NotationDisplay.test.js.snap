// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders the note given by the prop 1`] = `
<path
  d="M111.733 159.946C106.37 159.946,103 162.164,103 164.916C103 167.64,105.303 170.054,111.34 170.054C117.967 170.054,120.073 167.752,120.073 164.916C120.073 162.052,115.496 159.946,111.733 159.946M107.493 162.445C107.942 161.041,109.43 160.844,110.694 160.844C113.474 160.844,115.692 163.821,115.692 166.264C115.692 167.499,115.159 168.65,113.839 168.959C113.446 169.072,112.996 169.128,112.575 169.128C111.115 169.128,109.627 168.145,108.784 167.022C107.97 166.095,107.38 164.719,107.38 163.428C107.38 163.091,107.409 162.782,107.493 162.445"
  stroke="none"
/>
`;

exports[`it renders the note given by the prop 2`] = `
<path
  d="M111.733 159.946C106.37 159.946,103 162.164,103 164.916C103 167.64,105.303 170.054,111.34 170.054C117.967 170.054,120.073 167.752,120.073 164.916C120.073 162.052,115.496 159.946,111.733 159.946M107.493 162.445C107.942 161.041,109.43 160.844,110.694 160.844C113.474 160.844,115.692 163.821,115.692 166.264C115.692 167.499,115.159 168.65,113.839 168.959C113.446 169.072,112.996 169.128,112.575 169.128C111.115 169.128,109.627 168.145,108.784 167.022C107.97 166.095,107.38 164.719,107.38 163.428C107.38 163.091,107.409 162.782,107.493 162.445"
  stroke="none"
/>
`;
